src/terminal/adapter/adaptDispatch.cpp

```diff
@@ -394,60 +417,100 @@ bool AdaptDispatch::CursorRestoreState()
+        // of arbitrary widths at some point in the future.
+        const auto top = gsl::narrow_cast<short>(delta > 0 ? scrollRect.top : (scrollRect.top + absoluteDelta));
+        const auto height = gsl::narrow_cast<short>(scrollRect.height() - absoluteDelta);
+        const auto actualDelta = gsl::narrow_cast<short>(delta > 0 ? absoluteDelta : -absoluteDelta);
+        textBuffer.ScrollRows(top, height, actualDelta);
+        textBuffer.TriggerRedraw(Viewport::FromInclusive(scrollRect.to_small_rect()));
     }
-    else
+
+    // Rows revealed by the scroll are filled with standard erase attributes.
+    auto eraseRect = scrollRect;
+    eraseRect.top = delta > 0 ? scrollRect.top : (scrollRect.bottom - absoluteDelta);
+    eraseRect.bottom = eraseRect.top + absoluteDelta;
+    auto eraseAttributes = textBuffer.GetCurrentAttributes();
+    eraseAttributes.SetStandardErase();
+    _FillRect(textBuffer, eraseRect, L' ', eraseAttributes);
+
+    // Also reset the line rendition for the erased rows.
+    textBuffer.ResetLineRenditionRange(eraseRect.top, eraseRect.bottom);
+}
+
+// Routine Description:
+// - Scrolls an area of the buffer in a horizontal direction.
+// Arguments:
+// - textBuffer - Target buffer to be scrolled.
+// - fillRect - Area of the buffer that will be affected.
+// - delta - Distance to move (positive is right, negative is left).
+// Return Value:
+// - <none>
+void AdaptDispatch::_ScrollRectHorizontally(TextBuffer& textBuffer, const til::rect scrollRect, const int32_t delta)
+{
+    const auto absoluteDelta = std::min(std::abs(delta), scrollRect.width());
+    if (absoluteDelta < scrollRect.width())
     {
-        // Delete scrolls the affected region to the left, relying on the clipping rect to actually delete the characters.
-        THROW_IF_FAILED(ShortSub(coordDestination.X, distance, &coordDestination.X));
+        const auto left = gsl::narrow_cast<short>(delta > 0 ? scrollRect.left : (scrollRect.left + absoluteDelta));
+        const auto top = gsl::narrow_cast<short>(scrollRect.top);
+        const auto width = gsl::narrow_cast<short>(scrollRect.width() - absoluteDelta);
+        const auto height = gsl::narrow_cast<short>(scrollRect.height());
+        const auto actualDelta = gsl::narrow_cast<short>(delta > 0 ? absoluteDelta : -absoluteDelta);
+
+        const auto source = Viewport::FromDimensions({ left, top }, width, height);
+        const auto target = Viewport::Offset(source, { actualDelta, 0 });
+        const auto walkDirection = Viewport::DetermineWalkDirection(source, target);
+        auto sourcePos = source.GetWalkOrigin(walkDirection);
+        auto targetPos = target.GetWalkOrigin(walkDirection);
+        do
+        {
+            const auto data = OutputCell(*textBuffer.GetCellDataAt(sourcePos));
+            textBuffer.Write(OutputCellIterator({ &data, 1 }), targetPos);
+            source.WalkInBounds(sourcePos, walkDirection);
+        } while (target.WalkInBounds(targetPos, walkDirection));
```

Comment:
Coreect me if I'm wrong here - if we had a buffer like:

```
1234567890
```

And the cursor was on the `1`, and we inserted 5 characters, we'd call this with `(textBuffer, { 0, 0, 80, 1 }, 5)`


<hr>

Actually nevermind me. That works perfectly fine. I thought we'd start reading from the section we were writing from, so it'd look like 

```
12341234
```

or something, but nah this works fine.

Reply:
I think I just stole that code from the existing `_CopyRectangle` function:
https://github.com/microsoft/terminal/blob/ce6ce50e7e9eeb2d7377527d9512da9d43674bc0/src/host/output.cpp#L107-L111